src/pybind/mgr/crash/module.py

```diff
@@ -0,0 +1,248 @@
+
+    def handle_command(self, inbuf, command):
+        for cmd in self.COMMANDS:
+            if cmd['cmd'].startswith(command['prefix']):
+                handler = cmd['handler']
+                break
+        if handler is None:
+            return errno.EINVAL, '', 'unknown command %s' % command['prefix']
+
+        return handler(self, command, inbuf)
+
+    @staticmethod
+    def validate_crash_metadata(inbuf):
+        # raise any exceptions to caller
+        metadata = json.loads(inbuf)
+        if 'crash_id' not in metadata:
+            raise AttributeError("missing 'crash_id' field")
+        return metadata
+
+    @staticmethod
+    def time_from_string(timestr):
+        # drop the 'Z' timezone indication, it's always UTC
+        timestr = timestr.rstrip('Z')
+        return datetime.datetime.strptime(timestr, DATEFMT)
+
+    # command handlers
+
+    def do_info(self, cmd, inbuf):
+        crashid = cmd['id']
+        key = 'crash/%s' % crashid
+        val = self.get_store(key)
+        if not val:
+            return errno.EINVAL, '', 'crash info: %s not found' % crashid
+        return 0, val, ''
+
+    def do_post(self, cmd, inbuf):
+        try:
+            metadata = self.validate_crash_metadata(inbuf)
+        except Exception as e:
+            return errno.EINVAL, '', 'malformed crash metadata: %s' % e
+
+        crashid = metadata['crash_id']
+        key = 'crash/%s' % crashid
+        # repeated stores of same item are ignored silently
+        if not self.get_store(key):
+            self.set_store(key, inbuf)
+        return 0, '', ''
+
+    def do_ls(self, cmd, inbuf):
+        keys = []
+        for key in self.get_store_prefix('crash/').iterkeys():
+            keys.append(key.replace('crash/', ''))
+        return 0, '\n'.join(keys), ''
+
+    def do_rm(self, cmd, inbuf):
+        crashid = cmd['id']
+        key = 'crash/%s' % crashid
+        if self.get_store(key) is None:
+            return errno.EINVAL, '', 'crash rm: id %s not present' % crashid
+        self.set_store(key, None)       # removes key
+        return 0, '', ''
+
+    def do_prune(self, cmd, inbuf, testmeta=None):
+        now = datetime.datetime.utcnow()
+
+        keep = cmd['keep']
+        try:
+            keep = int(keep)
+        except ValueError:
+            return errno.EINVAL, '', 'keep argument must be integer'
+
+        keeptime = datetime.timedelta(days=keep)
+
+        # testability
+        if testmeta is not None:
+            iterator = testmeta
+            test_removed = list()
+        else:
+            iterator = self.get_store_prefix('crash/')
+
+        for key, meta in iterator.iteritems():
+            meta = json.loads(meta)
+            stamp = self.time_from_string(meta['timestamp'])
+            if stamp <= now - keeptime:
+                if testmeta:
+                    test_removed.append('rm(%s)' % key)
+                else:
+                    self.do_rm(cmd, inbuf)
```

Comment:
I don't think this works -- it's passing through the same `cmd` that came into do_prune.

I'd just call set_store(key, None) directly here.

Reply:
sigh, you're correct of course.  I swear I tested this once.  Will fix and verify actual operation somehow.  Where should I put this sort of actual test?